### DISCUSSION TITLE
Why exactly are API levels < 26 not supported?

### DISCUSSION THREAD

**Discussion by bddckr** [mannodermaus/android-junit5#332] [Q&A] [answered]
While the Readme explains

>Because the framework is built on Java 8 from the ground up, these instrumentation tests will only run on devices running Android 8.0 (API 26) or newer – older phones will skip the execution of these tests completely, marking them as "ignored".

I'm wondering what exactly the "issue" is to support earlier API levels.

In a very crude early test, I could execute some tests on API level 25 (my min SDK) without issues. Perhaps it's because I've enabled [AGP's desugaring support](https://developer.android.com/studio/write/java8-support#library-desugaring)?

If it boils down to _some_ JUnit 5 functionality not being available, I'd like to request this check, perhaps via some explicit opt-in flag in the Gradle plugin this project provides? 👀 
https://github.com/mannodermaus/android-junit5/blob/38ef72c06afe2dc23e94bf7d32cff5a41992b52c/instrumentation/runner/src/main/kotlin/de/mannodermaus/junit5/internal/runners/JUnit5RunnerFactory.kt#L21-L25

**Comment by mannodermaus** [comment, marked as answer]
The reason why you cannot run JUnit Jupiter tests below API 26 is that JUnit 5 uses several APIs that aren't available on devices shipping with Java 7 or older, and the fact that these APIs aren't completely covered by core library desugaring either. For example, the upcoming release candidate for JUnit 5.11 will fail right on discovery because [`Class#getTypeName`](https://developer.android.com/reference/java/lang/Class?hl=en#getTypeName()) cannot be found on older devices. Because of this, it's not a matter of excluding only _some_ functionality since the main entry point fails immediately.

That being said, I'd love to see this sample project of yours if you can share it!

**Reply by bddckr** [reply]
> because [Class#getTypeName](https://developer.android.com/reference/java/lang/Class?hl=en#getTypeName()) cannot be found on older devices

Ah, too bad. That alone makes this whole path futile. My example is so basic, it probably just didn't run into any of those issues just yet.

Thanks for the quick reply! I'll just consider ignoring API level 25 for my project.
